image-squashfs.c

```diff
@@ -40,11 +46,38 @@ static int squash_generate(struct image *image)
 	else
 		snprintf(compression, sizeof(compression), "-comp %s", comp_setup);
 
-	return systemp(image, "%s '%s' '%s' -b %u -noappend %s %s",
-			get_opt("mksquashfs"),
-			mountpath(image), /* source dir */
-			imageoutfile(image), /* destination file */
-			block_size, compression, extraargs);
+	ret = systemp(image, "%s '%s' '%s' -b %u -noappend %s %s",
+		      get_opt("mksquashfs"),
+		      mountpath(image), /* source dir */
+		      imageoutfile(image), /* destination file */
+		      block_size, compression, extraargs);
+	if (ret)
+		return ret;
+	ret = stat(imageoutfile(image), &sb);
+	if (ret) {
+		ret = -errno;
+		image_error(image, "stat(%s) failed: %s\n", imageoutfile(image), strerror(errno));
+		return ret;
+	}
+	file_size = sb.st_size;
+
+	if (!image->size)
+		image->size = file_size;
+
+	if (file_size > image->size) {
+		image_error(image, "generated image %s is larger than given image size (%llu v %llu)\n",
+			    imageoutfile(image), file_size, image->size);
+		return -E2BIG;
+	} else if (file_size < image->size) {
+		ret = truncate(imageoutfile(image), image->size);
```

Comment:
Fair enough. I've pushed a new version. I've chosen to always set image->size to the actual file size and not leave it at some possibly bigger value, since consumers (e.g. image-hd) that rely on some particular size would probably just set the partition size explicitly.

Reply:
Makes sense.